Mark cookies secure

gherceg opened the pull request:
<!--- Provide a link to the ticket or document which prompted this change -->
[Mark cookies secure](https://dimagi-dev.atlassian.net/browse/SAAS-11314)
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
As the ticket explains, we currently do not mark cookies as secure. Marking cookies secure ensures they are only sent over an HTTPS connection. 
##### FEATURE FLAG
<!--- If this is specific to a feature flag, which one? -->

##### RISK ASSESSMENT / QA PLAN
<!-- Does this need QA before or after merge? Link QA ticket. -->
[QA Ticket](https://dimagi-dev.atlassian.net/browse/QA-2099)
##### PRODUCT DESCRIPTION
<!--- For non-invisible changes, describe user-facing effects. -->

gherceg committed:
secure cookies in python

gherceg committed:
secure cookies in javascript

orangejenny commented:
Is this going to affect local development, since most/all of use http there?

gherceg commented:
> Is this going to affect local development, since most/all of us use http there?

Good question @orangejenny. The [mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) state this "A cookie with the Secure attribute is sent to the server only with an encrypted request over the HTTPS protocol, never with unsecured HTTP (except on localhost)". Knowing that, is there still potential for someone's local workflow to be affected? I know some use 127.0.0.1 instead of localhost and I'm not sure if this makes a difference in this context. My assumption is no.

orangejenny commented:
@gherceg Great, thank you for talking through that.

gherceg committed:
move secure value to settings.py and add to python files

gherceg committed:
pass settings.SECURE_COOKIES value into js files that need it via python and html files

gherceg committed:
fix js errors

gherceg commented:
@dannyroberts @orangejenny sorry for the large commit. Pulling the secure cookies values into settings as `SECURE_COOKIES` resulted in a lot of necessary changes to make this value accessible in javascript. I followed precedent for the most part. The `BasePageView` tends to cover most cases for passing in the `secure_cookies` value in the view's context.

orangejenny commented on corehq/apps/hqwebapp/templates/hqwebapp/base.html:
```diff
@@ -359,6 +359,7 @@
     {% initial_page_data 'show_mobile_ux_warning' show_mobile_ux_warning %}
     {% initial_page_data 'toggles_dict' toggles_dict %}
     {% initial_page_data 'previews_dict' previews_dict %}
+    {% initial_page_data 'secure_cookies' secure_cookies %}
```
Including this should mean you don't need to add it anywhere else, since virtually all HQ views ultimately derive from this template. But double-check the app preview view, that might be one of the few that doesn't.

gherceg replied:
Ah right. I had this revelation on the Python side but not on the HTML side. Will make that change for anything that inherits from `base.html`. Thanks

orangejenny replied:
You might also consider adding this value somewhere in [context_processors](https://github.com/dimagi/commcare-hq/blob/master/corehq/util/context_processors.py) rather than adding it to individual views. It's definitely a tradeoff, since that'll add a little more bloat to all requests.

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/rollout_modal.js:
```diff
@@ -14,10 +14,10 @@ hqDefine("hqwebapp/js/rollout_modal", [
     googleAnalytics,
     kissmetricsAnalytics
 ) {
+    var secureCookies = initialPageData.get('secure_cookies');
```
You can kill this whole file (and `hqwebapp/rollout_modal.html`) if you like, it's not being used. Also fine to leave it alone if you don't want to increase the scope of this PR.

gherceg replied:
I can make another PR after this has been merged to handle that 👍

orangejenny commented on corehq/apps/hqwebapp/templatetags/hq_shared_tags.py:
```diff
@@ -685,6 +685,7 @@ def javascript_libraries(context, **kwargs):
         'analytics': kwargs.pop('analytics', False),
         'hq': kwargs.pop('hq', False),
         'helpers': kwargs.pop('helpers', False),
+        'secure_cookies': settings.SECURE_COOKIES,
```
What's this change for?

gherceg replied:
the `core_libraries.html` file includes the `hqwebapp/js/components/pagination.js` script and that was the only spot I saw `core_libraries.html` used so I figured that's where the context would get setup for it.

orangejenny replied:
You're right that this is the context for `core_libraries.html`, but since you're not adding anything to `core_libraries.html` itself, I don't think this is doing anything. Since this is a template tag, `core_libraries.html` isn't included directly in other templates, it gets inserted by the `javascript_libraries` tag - [this](https://github.com/dimagi/commcare-hq/blob/f8f29068699aeadd0d8b0cf488acec5150926b0e/corehq/apps/hqwebapp/templates/hqwebapp/base.html#L136) is the main place it's used.

I didn't notice earlier that `pagination.js` needs `secure_cookies`. Because of that `javascript_libraries` tag in `hqwebapp/base.html`, `pagination.js` gets included in virtually every page on HQ, so `secure_cookies` also needs to be available basically everywhere. Considering that, I now more strongly think `secure_cookies` should go into context processors instead of being passed into individual views.

gherceg committed:
revert changes on html templates that inherit from base.html

gherceg committed:
pass secure_cookies in context_processors to every view request

gherceg committed:
fix loading initial page data issues

- reorder dependencies in core_libraries.html to include initial_page_data.js before pagination.js
- defer fetching 'secure_cookies' from intialPageData in js code to allow page time to load

orangejenny reviewed (approved):
Updates look good.

gherceg committed:
Merge branch 'master' into gh/secure-cookies

gherceg committed:
fix order of imports test failure

gherceg committed:
do not add secure attribute on cookie get methods

gherceg commented:
Requested QA.

dannyroberts commented on corehq/apps/app_manager/views/view_generic.py:
```diff
@@ -1,3 +1,4 @@
+from django.conf import settings
```
nit: looks like an extraneous change

gherceg committed:
revert changes to view_generic.py

gherceg committed:
Merge branch 'master' into gh/secure-cookies

millerdev commented:
:tada:
